### PR TITLE
[ASL] Rework binary operation effects on symbolic constraints

### PR DESCRIPTION
**Pull request by HadrienRenaud**
- [x] Reorganise the code for binary-operation checking: it is now more explicit and shorter.
- [x] Improve DIV results by considering it decreasing on its second argument (see bug report later)
- [x] Add test for this
- [x] Merge static intervals after divisions using a diet
- [x] Implement `constraint_binop` for `MOD` and remove all usages of this strange default to `UnConstrained`
- [x] Fix DIV implementation: DIV is not decreasing for negative numerators.
- [x] Fix MUL, SHR, SHL, POW implementation: they are not increasing if one operand is negative.

---

### Bug report

With the following test, we can see that the results of divisions on constraints are defaulting to the unconstrained integer. This is unwanted behaviour.

```
func main() =>integer
begin
  let a : integer {10} = 10;
  let b : integer {10..20} = 10;
  var c = b DIV a;
  var d = a DIV b;
  c = 1.0; // will error so that the type is printed
  d = 1.0; // will error so that the type is printed
  return 0;
end
```

Previous output for this test:
```
❯ aslref tests-asl/1008.asl --type-check-warn
File tests-asl/1008.asl, line 7, characters 2 to 3:
ASL Typing error: a subtype of integer {1..2} was expected, provided real.
File tests-asl/1008.asl, line 8, characters 2 to 3:
ASL Typing error: a subtype of integer was expected, provided real.
```

Current output for this test:
```
❯ dune exec aslref -- tests-asl/1008.asl --type-check-warn
File tests-asl/1008.asl, line 7, characters 2 to 3:
ASL Typing error: a subtype of integer {1..2} was expected, provided real.
File tests-asl/1008.asl, line 8, characters 2 to 3:
ASL Typing error: a subtype of integer {1} was expected, provided real.
```

### CODE REVIEW

**Comment by maranget** [comment]
Hi @HadrienRenaud, I am not sure that the warnings are always useful : "trop de warning tue le warning". Maybe we should consider various levels of verbosity controlled by some command-line option?

File: asllib/ASTUtils.ml [hunk 1]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
     [] li1
   |> List.rev
 
+let list_flat_cross f li1 li2 =
+  List.fold_left
+    (fun xys x ->
+      List.fold_left (fun xys' y -> List.rev_append (f x y) xys') xys li2)
+    [] li1
+  |> List.rev
+
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:603]: (**
Tests whether the value of binary operation [op] increases along with its left-hand-side operand when all operands are non-negative. That is, if [is_left_increasing op] is true then for every two expressions [x1 op y] and [x2 op y] such that [x1 >= 0 /\ x2 >= 0 /\ y >=0] and [x1 <= x2] then [x1 op y <= x2 op y].
*)

File: asllib/ASTUtils.ml [hunk 2]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
 let is_left_increasing = function
-  | MUL | DIV | DIVRM | SHL | SHR | POW | PLUS | MINUS -> true
-  | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MUL | SHL | SHR | POW | PLUS | MINUS -> true
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:605]: If you agree with the definition given by my comment above, shouldn't DIV/DIVRM be true?

File: asllib/ASTUtils.ml [hunk 3]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
+  | DIV | DIVRM | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:606]: Shouldn't the result for MOD be false, rather than a failed assertion?
It looks like the function either returns true or fails.

File: asllib/ASTUtils.ml [hunk 4]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_increasing = function
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:610]: (**
Tests whether the value of binary operation [op] increases along with its right-hand-side operand when all operands are non-negative. That is, if [is_right_increasing op] is true then for every two expressions [x op y1] and [x op y2] such that [y1 >= 0 /\ y2 >= 0 /\ x >=0]  and [y1 <= y2] then [x op y1 <= x op y2].
*)

File: asllib/ASTUtils.ml [hunk 5]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
   | MUL | SHL | SHR | POW | PLUS -> true
-  | DIV | DIVRM | MINUS -> false
-  | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MINUS -> false
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:612]: shouldn't SHR be false? 8 >> 1 is greater than 8 >> 2

File: asllib/Typing.ml [hunk 6]
```diff
@@ -618,7 +618,8 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
 
   let binop_is_exploding = function
     | MUL | SHL | POW -> true
-    | PLUS | DIV | MINUS | MOD | SHR | DIVRM -> false
+    | DIV | DIVRM -> (* general case loses too much precision *) true
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:621]: I don't understand the comment. Perhaps add a comment explaining this function?

File: asllib/ASTUtils.ml [hunk 7]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
+  | DIV | DIVRM | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_decreasing = function
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:617]: (**
Tests whether the value of binary operation [op] decreases along with its right-hand-side operand when all operands are non-negative. That is, if [is_right_decreasing op] is true then for every two expressions [x op y1] and [x op y2] such that [y1 >= 0 /\ y2 >= 0 /\ x >=0] and [y1 <= y2] then [x op y1 <= x op y2].
*)

Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:617]: Do we even need this function? It's the boolean negation of is_right_incvreasing.

File: asllib/ASTUtils.ml [hunk 8]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
-    | Constraint_Exact e | Constraint_Range (_, e) ->
-        Constraint_Range (zero_expr, e)
+    | Constraint_Exact e | Constraint_Range (_, e) -> range zero_expr e
+  in
+  let constraint_divisions op c1 c2 =
+    match (c1, c2) with
+    | Constraint_Exact e1, Constraint_Exact e2 -> [ exact (binop op e1 e2) ]
+    | Constraint_Range (e11, e12), Constraint_Exact e2 ->
+        [ range (binop op e11 e2) (binop op e12 e2) ]
+    | Constraint_Exact e1, Constraint_Range (_e21, e22) ->
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:658]: Is this correct?
For example `10 DIV 1..10` = `1, 2, 5, 10`, not `[10..1, 1..10]`

File: asllib/ASTUtils.ml [hunk 9]
```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
...
+        (* {a} DIV {b..c} == {a .. (a DIV c), (a DIV c) .. a} *)
+        [ range e1 (binop op e1 e22); range (binop op e1 e22) e1 ]
+    | Constraint_Range (e11, e12), Constraint_Range (_e21, _e22) ->
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:661]: Likewise, I'm not sure the result is a fixed number of ranges.

File: asllib/ASTUtils.ml [hunk 10]
```diff
@@ -108,6 +108,16 @@ let list_concat_map f l =
   in
   aux f [] l
 
+let list_take =
```
Comment (Roman-Manevich) [inline asllib/ASTUtils.ml:111]: How about renaming to list_prefix?

Comment (Roman-Manevich) [reply asllib/ASTUtils.ml:111]: Okay, but how about a documentation comment saying that it returns a prefix?

File: asllib/StaticOperations.ml [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+  | SHR -> list_flat_cross constraint_shr cs1 cs2
+  | MOD -> List.map constraint_mod cs2
+  | SHL -> list_flat_cross constraint_shl cs1 cs2
+  | POW -> list_flat_cross constraint_pow cs1 cs2
+  | AND | BAND | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ | NEQ | OR
+  | RDIV ->
+      assert false
+(* End *)
+
+(* Begin FilterReduceConstraintDiv *)
```
Comment (Roman-Manevich) [inline asllib/StaticOperations.ml:148]: It's getting hard to follow the logic here. I think some comments could help.

**Review by Roman-Manevich — CHANGES_REQUESTED** [review]
